kotlinx-coroutines-core/common/src/selects/Select.kt

```diff
@@ -52,606 +97,696 @@ public interface SelectBuilder<in R> {
- * corresponding non-suspending version that can be used with a regular `when` expression to select one
- * of the alternatives or to perform the default (`else`) action if none of them can be immediately selected.
- *
- * ### List of supported select methods
- *
- * | **Receiver**     | **Suspending function**                           | **Select clause**
- * | ---------------- | ---------------------------------------------     | -----------------------------------------------------
- * | [Job]            | [join][Job.join]                                  | [onJoin][Job.onJoin]
- * | [Deferred]       | [await][Deferred.await]                           | [onAwait][Deferred.onAwait]
- * | [SendChannel]    | [send][SendChannel.send]                          | [onSend][SendChannel.onSend]
- * | [ReceiveChannel] | [receive][ReceiveChannel.receive]                 | [onReceive][ReceiveChannel.onReceive]
- * | [ReceiveChannel] | [receiveCatching][ReceiveChannel.receiveCatching] | [onReceiveCatching][ReceiveChannel.onReceiveCatching]
- * | none             | [delay]                                           | [onTimeout][SelectBuilder.onTimeout]
- *
- * This suspending function is cancellable. If the [Job] of the current coroutine is cancelled or completed while this
- * function is suspended, this function immediately resumes with [CancellationException].
- * There is a **prompt cancellation guarantee**. If the job was cancelled while this function was
- * suspended, it will not resume successfully. See [suspendCancellableCoroutine] documentation for low-level details.
- *
- * Note that this function does not check for cancellation when it is not suspended.
- * Use [yield] or [CoroutineScope.isActive] to periodically check for cancellation in tight loops if needed.
- */
-public suspend inline fun <R> select(crossinline builder: SelectBuilder<R>.() -> Unit): R {
-    contract {
-        callsInPlace(builder, InvocationKind.EXACTLY_ONCE)
-    }
-    return suspendCoroutineUninterceptedOrReturn { uCont ->
-        val scope = SelectBuilderImpl(uCont)
-        try {
-            builder(scope)
-        } catch (e: Throwable) {
-            scope.handleBuilderException(e)
+    private val inRegistrationPhase
+        get() = state.value.let {
+            it === STATE_REG || it is List<*>
         }
-        scope.getResult()
-    }
-}
+    /**
+     * Returns `true` if this `select` is already selected or cancelled;
+     * thus, other parties are bound to fail when making a rendezvous with it.
+     */
+    private val isSelected
+        get() = state.value is ClauseData<*>
+    /**
+     * Returns `true` if this `select` is cancelled.
+     */
+    private val isCancelled
+        get() = state.value is Cancelled
 
+    /**
+     * List of clauses waiting on this `select` instance.
+     */
+    private var clauses: MutableList<ClauseData<R>>? = ArrayList(2)
```

Comment:
Unfortunately, `emptyList()` is not an instance of `MutableList`. Thus, I would keep it as is.

Reply:
Then I suggest just calling `clear()` instead of nulling out the slot.

For now, `null` doesn't really prevent anything except making code much harder to read -- somewhere `!!` is used, and `?` in other places, making the reasoning really hard, I've stumbled across it once again

If users abuse API hard (e.g. by storing `select` instance in a field), they will get caught by our asserts, and if it's a race -- all bets are off anyway

Reply:
Then there is another problem --  the `findClause` function iterates over the list of clauses and can be invoked concurrently with `clear()`, which leads to throwing a `ConcurrentModificationException`. Thus, the best way is to replace the existing mutable list with either `null` or `emptyList()`.

Reply:
Then probably setting it to empty mutable list with a corresponding commentary about why it is applicable should work

Reply:
I have improved the documentation.